larpandoracontent/LArObjects/LArDiscreteProbabilityVector.cc

```diff
@@ -0,0 +1,240 @@
+        const float deltaX(resamplingPoints.at(iSample+1)-xResampled);
+        const float cumulativeDatumResampled(discreteProbabilityVector.EvaluateCumulativeProbability(xResampled));
+        const float densityDatumResampled((cumulativeDatumResampled-prevCumulativeData)/(m_useWidths ? deltaX : 1.f));
+        resampledProbabilityData.emplace_back(DiscreteProbabilityVector::DiscreteProbabilityDatum(xResampled, 
+            densityDatumResampled, cumulativeDatumResampled, deltaX));
+        prevCumulativeData = cumulativeDatumResampled;
+    }
+
+    const float xResampled(resamplingPoints.back());
+    const float deltaX(m_xUpperBound-xResampled);
+    const float cumulativeDatumResampled(discreteProbabilityVector.EvaluateCumulativeProbability(xResampled));
+    const float densityDatumResampled((cumulativeDatumResampled-prevCumulativeData)/(m_useWidths ? deltaX : 1.f));
+    resampledProbabilityData.emplace_back(DiscreteProbabilityVector::DiscreteProbabilityDatum(xResampled, densityDatumResampled, 
+        cumulativeDatumResampled, deltaX));
+
+    return resampledProbabilityData;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+DiscreteProbabilityVector::DiscreteProbabilityData DiscreteProbabilityVector::RandomiseDiscreteProbabilityData(
+   DiscreteProbabilityVector const &discreteProbabilityVector, std::mt19937 &randomNumberGenerator) const
+{
+    DiscreteProbabilityData randomisedProbabilityData;
+
+    std::vector<size_t> randomisedElements(discreteProbabilityVector.GetSize());
+    std::iota (std::begin(randomisedElements), std::end(randomisedElements), 0);
+    std::shuffle(std::begin(randomisedElements), std::end(randomisedElements), randomNumberGenerator);
+
+    float xPos(discreteProbabilityVector.GetX(0));
+    float cumulativeProbability(0.f);
+    for (size_t iElement = 0; iElement < discreteProbabilityVector.GetSize(); ++iElement)
+    {
+        const size_t randomElementIndex(randomisedElements.at(iElement));
+        const float deltaX(discreteProbabilityVector.GetWidth(randomElementIndex));
+        const float probabilityDensity(discreteProbabilityVector.GetProbabilityDensity(randomElementIndex));
+        cumulativeProbability+=probabilityDensity*(m_useWidths ? deltaX : 1.f);
+        randomisedProbabilityData.emplace_back(DiscreteProbabilityVector::DiscreteProbabilityDatum(xPos, probabilityDensity, 
+            cumulativeProbability, deltaX));
+        xPos+=deltaX;
+    }
+
+    return randomisedProbabilityData;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+template <typename TX, typename TY>
+bool DiscreteProbabilityVector::SortInputDataByX(InputDatum<TX, TY> lhs, InputDatum<TX, TY> rhs)
+{
+    float deltaX(static_cast<float>(rhs.first) - static_cast<float>(lhs.first));
```

Comment:
`const`?

Reply:
Fixed. 
Thank you, my const guardian angel.